16_permutations/solution/permutations-solution.js

```diff
@@ -0,0 +1,19 @@
+const permutations = function (original, currentPermutations) {
+  if (original.length === 1) return original;
+
+  const perms = currentPermutations
+    ? currentPermutations
+    : original.map((el) => [el]);
+
+  const newPerms = [];
+  perms.forEach((el) => {
+    const missing = original.filter((item) => !el.includes(item));
+    missing.forEach((itemMissing) => newPerms.push([...el, itemMissing]));
+  });
+
+  if (newPerms.every((el) => el.length === original.length)) return newPerms;
+  return permutations(original, newPerms);
+};
```

Comment:
I'll be honest, I had quite a bit of trouble understanding this solution as a whole. Not because the method itself is bad, I actually think it's really elegant. My difficult primarily came from the naming of things, as several of the things aren't really named very accurately. Hence it threw me off as I was expecting some things to represent one thing, but what they actually represented (and thus how they were ultimately used) was something very different.

Once I clocked what they actually represented, the approach made total sense.

`currentPermutations` to me, indicates something that contains all actual permutations at this point in time. So for an input array of `[1, 2, 3, 4]`, a `currentPermutations` array containing `[[1, 2, 3, 4], [1, 2, 4, 3]]` makes sense to me. Each element is a permutation, and the whole array contains only the permutations currently found.

Instead, what you're doing with each recursive call, is building up an array of **partial permutations**. For example, `[1]` and `[1, 4]` are both partial permutations of `[1, 2, 3, 4]`, but the `currentPermutations` naming implies they're full permutations.

The below is a suggestion, but implements the above comment's changes as well as different naming. Checking `.every` for the end base case is fine, but by nature of this approach building up partial permutations one element at a time, all elements will be the same length as each other at every recursive call. Also no strong opinion on `perm`/`permutation`, just personally prefer the full. Let me know what you think.

```js
const permutations = function (original, partialPermutations = original.map((num) => [num])) {
  if (original.length === 0) return [];
  if (original.length === 1) return [original];

  const newPartialPermutations = [];
  partialPermutations.forEach((partialPermutation) => {
    const missingNums = original.filter((num) => !partialPermutation.includes(num));
    missingNums.forEach((missingNum) => {
      newPartialPermutations.push([...partialPermutation, missingNum]);
    });
  });

  if (newPartialPermutations[0].length === original.length) {
    return newPartialPermutations;
  }
  return permutations(original, newPartialPermutations);
};
```

Reply:
I should also note that the solution will not work if the input array has duplicates.
You may prefer to not include this additional complexity, so if that's the case, the instructions should also make clear that the learner will not need to account for duplicate numbers.

Reply:
> I'll be honest, I had quite a bit of trouble understanding this solution as a whole. Not because the method itself is bad, I actually think it's really elegant (like, *very*). My difficulty primarily came from the naming of things, as several of the things aren't really named very accurately. Hence it threw me off as I was expecting some things to represent one thing, but what they actually represented (and thus how they were ultimately used) was something very different.
> 
> Once I clocked what they actually represented, the approach made total sense.
> 
> `currentPermutations` to me, indicates something that contains all actual permutations at this point in time. So for an input array of `[1, 2, 3, 4]`, a `currentPermutations` array containing `[[1, 2, 3, 4], [1, 2, 4, 3]]` makes sense to me. Each element is a permutation, and the whole array contains only the permutations currently found.
> 
> Instead, what you're doing with each recursive call, is building up an array of **partial permutations**. For example, `[1]` and `[1, 4]` are both partial permutations of `[1, 2, 3, 4]`, but the `currentPermutations` naming implies they're full permutations.
> 
> The below is a suggestion, but implements the above comment's changes as well as different naming. Checking `.every` for the end base case is fine, but by nature of this approach building up partial permutations one element at a time, all elements will be the same length as each other at every recursive call. Also no strong opinion on `perm`/`permutation`, just personally prefer the full. Let me know what you think.
> 
> ```js
> const permutations = function (original, partialPermutations = original.map((num) => [num])) {
>   if (original.length === 0) return [];
>   if (original.length === 1) return [original];
> 
>   const newPartialPermutations = [];
>   partialPermutations.forEach((partialPermutation) => {
>     const missingNums = original.filter((num) => !partialPermutation.includes(num));
>     missingNums.forEach((missingNum) => {
>       newPartialPermutations.push([...partialPermutation, missingNum]);
>     });
>   });
> 
>   if (newPartialPermutations[0].length === original.length) {
>     return newPartialPermutations;
>   }
>   return permutations(original, newPartialPermutations);
> };
> ```
> 
> 

This def makes the code a lot more readable. Misnaming a single variable can lead to so much confusion, I'll certainly take a moment to take this in as probably misnamed variables are harder to understand than a variable that has a random meaningless name lol.

I agree that `partialPermutations` makes significantly more sense than `currentPermutations`, thank you for pointing this out.

Reply:
Ok, I've made the changes
- improved variable naming
- clarified that the same integer will not be given twice in the README
- added example and tests for empty array
- refactored to use default parameter